File: solutions/kotlin/prime-factors/2/src/main/kotlin/PrimeFactors.kt
object PrimeFactorCalculator {

    fun primeFactors(number: Int): List<Int> {
        return primeFactors(number.toLong()).map { it.toInt() }
    }

    fun primeFactors(number: Long): List<Long> {
        val factors = mutableListOf<Long>()
        var remainingNumber = number

        while (remainingNumber % 2 == 0L) {
            factors.add(2L)
            remainingNumber /= 2
        }

        var divisor = 3L
        while (divisor * divisor <= remainingNumber) {
            while (remainingNumber % divisor == 0L) {
                factors.add(divisor)
                remainingNumber /= divisor
            }
            divisor += 2
        }

        if (remainingNumber > 1) {
            factors.add(remainingNumber)
        }

        return factors
    }
}

File: solutions/go/prime-factors/1/prime_factors.go
package prime


func Factors(n int64) []int64 {
	factors := make([]int64, 0)


	for n%2 == 0 {
		factors = append(factors, 2)
		n /= 2
	}
	
	var divisor int64 = 3
	for divisor * divisor <= n {
		for n%divisor == 0 {
			factors = append(factors, divisor)
			n /= divisor
		}
		divisor += 2
	}
	if n > 1 {
		factors = append(factors, n)
	}
	return factors
}